### DISCUSSION TITLE
Monthly development updates

### DISCUSSION THREAD

**Discussion by amercader** [ckan/ckan#8213] [General]
## Context

As part of a new initiative started by Link Digital, @wardi, @amercader and @smotornyuk will be funded to work on the CKAN project, on what is known as "core work". That is working in the main CKAN repository and the supported extensions to develop new features, reduce tech debt and improve the developer and maintainer experience. You can find more details about it in this blog post:

https://ckan.org/blog/empowering-open-data-and-beyond-major-initiative-to-boost-ckan-development

We aim to be transparent about what we spend our time on, so we will be providing monthly updates in this discussion (we might change the frequency and publication place if necessary). Feel free to add your comments below to ask questions or know more about particular points (please use a thread below each month rather than the main thread)

**Comment by amercader** [comment]
## April 2024

This was our first month working on CKAN maintenance, so we spent quite a lot of time brainstorming and discussing what we could focus on. As you can imagine it was a big list, so we identified 4 key areas that we will focus on during the coming months.

1. [*CKAN 2.11 release*](https://github.com/ckan/ckan/milestone/15): there are some major features like [Table Designer](https://github.com/ckan/ckan/pull/7882) in the review pipeline, but also major refactorings and improvements to make CKAN more secure and performant. We've been making progress in many pending reviews to get as many of them merged as possible
2. *Developer and Maintainer experience*: these involve improvements and new tools around the main CKAN repo and the Docker images to provide a better experience when working on CKAN, both from a developer point of view (like enabling [VSCode remote debugging](https://github.com/ckan/ckan-docker-base/pull/50)) and from the maintainers one (like [automating backports](https://github.com/ckan/ckan/pull/8191)) Check out this [project](https://github.com/orgs/ckan/projects/7/views/1) to follow along other work being done in that front.
3. *DCAT v3*: CKAN is all about supporting common standards, and a good, robust DCAT support has always been an important goal. [An initial plan has been drawn](https://github.com/ckan/ckanext-dcat/issues/271) to build support for the new DCAT version, which involves work in the current ckanext-dcat version to align it with modern CKAN development patterns, before implementing DCAT-AP 3 and DCAT-US 3 support. Your feedback is more than welcome in the linked issue, and you can track progress in this [project](https://github.com/orgs/ckan/projects/5/views/1).
4. *File management*: There is a lot of room for improvement in how files are managed in CKAN. Think better support for large files, reducing CKAN load when accessing data, features like resumable uploads etc. The actual work plan is not yet defined, but @wardi has put down some initial details in https://github.com/ckan/ckan/issues/8197, and we will build on top of great work already done by @smotornyuk .


Apart from these main areas we of course worked on regular maintenance tasks like reviewing incoming pull requests and issues, drafting advisories for security issues, answering support requests etc.

**Reply by bjornhagstrom** [reply]
Great new that you are making DCAT-AP a priority. I checked the progress of that activity and I am thinking about support for national profiles of DCAT-AP (other then teh US that I guess is america?).Several european countries have national profiles, I myself is active in Sweden and know of a few home grown extentions that support DCAT-AP-SE. The activity you name "Scheming support", is that support for national profiles (or support for the possibility to use/define/support other profiles)?

**Reply by amercader** [reply]
@bjornhagstrom thanks for your comment. The [Scheming support](https://github.com/ckan/ckanext-dcat/pull/281) pull request includes preliminary work to integrate ckanext-dcat closer with the standard way site maintainers extend and customize metadata schemas with ckanext-scheming. Up until now ckanext-dcat focused on the mapping between CKAN and metadata and the DCAT based serializations that are mandated by the different DCAT-AP profiles, leaving to each site implementer to connect this to how datasets were created and managed. The goal with this work is to make transparent the connection between the fields defined in the CKAN schema and their DCAT representations, so for instance providing a base schema that implements all properties for dcat:Datasets and dcat:Distributions, taking care of things like the form UI for data collection (see e.g. an example below for dcat:contactPoint), translating it to the RDF serialization, loading codelists if necessary etc. There is also work to better support extensions like HVD, GeoDCAT-AP, etc
So to answer your question, yes, we definitely want to make easier to create and maintain national (and other) profiles and integrate the DCAT specific logic with other CKAN functionality.
Hope this makes sense.

![Screenshot 2024-05-31 at 11-26-05 Create Dataset - CKAN](https://github.com/ckan/ckan/assets/200230/33940d05-2a83-4c57-acc1-89d4a5f9d821)

**Reply by bjornhagstrom** [reply]
Thank you, that sounds great!

**Comment by amercader** [comment]
## May 2024

May really flew by really fast but we managed to make a lot of progress on several fronts:

* The biggest pending pull requests for the upcoming _2.11 release_ have been reviewed, polished and merged. [Table Designer](https://github.com/ckan/ckan/pull/7882), [a new Sessions backend](https://github.com/ckan/ckan/pull/7893), [Improved validation of id fields](https://github.com/ckan/ckan/pull/8069) and more. We also started working on integrating with Snyk to incorporate security upgrades more regularly. The new dev-v2.11 branch release is created and soon we'll start the formal release process, for which we'll ask the community for help, stay tuned!
* @wardi has been busy making lots of parts of CKAN faster: [startup time](https://github.com/ckan/ckan/pull/8221), [CLI](https://github.com/ckan/ckan/pull/8223), [template snippets](https://github.com/ckan/ckan/pull/8226) and @kowh-ai is working to improve our _Docker_ images to make them more performant, secure and easier to extend
* Lots of movement on the _DCAT_ front as well, initial [scheming support](https://github.com/ckan/ckanext-dcat/pull/281) is almost ready to go, plus good discussions around support for extensions like [HVD](https://github.com/ckan/ckanext-dcat/pull/284) and GeoDCAT-AP taking place

**Reply by wardi** [reply]
Approximate performance improvements:
 - cli/start-up time 20% faster
 - cli help commands 100% faster
 - page rendering up to 20% faster

**Comment by amercader** [comment]
## June 2024

Steady progress on the same fronts as the previous month: CKAN 2.11, Performance and DCAT

* We presented the current initiative and some of the work done so far in this month's [CKAN Monthly Live](https://www.youtube.com/watch?v=HKIMLwXG5pk), check it out if you missed it.
* There was a big push on the CKAN 2.11 release front, with work on improving our different Docker images and provide 2.11 images, which was required to start adding support for 2.11 in the [most popular extensions](https://github.com/ckan/ckan/issues/8291). Perhaps you can help test support on one of the pending ones?
* @smotornyuk created some [outstanding documentation](https://datashades.github.io/ckanext-files/) for the new ckanext-files extension that he has been working on. Make sure to check it out if you are interested in the all the new possibilities that it offers for files management.
* @wardi continues to improve performance and storage, this time at the database level with autogenerated alembic migrations, cleaning up duplicated foreign keys and indexes (https://github.com/ckan/ckan/pull/8239), and removing the groups and package extras tables (smaller DB, faster updates)

**Comment by kowh-ai** [comment]
## July 2024

* Lots of our efforts this month have been around preparing the upcoming CKAN 2.11 release (scheduled next ~~14th~~ 21st August if all goes well) while at the same time improving the release process and monitoring to have more robust releases and increase their frequency

   - We [tested all official extensions](https://github.com/ckan/ckan/issues/8291) and updated them for 2.11 support (that's 13 of them!), and likewise we tested the [different Python versions supported](https://github.com/ckan/ckan-python-monitor), making all necessary changes to ensure compatibility
   -  We set up Snyk to monitor the master and release branches, and did all the upgrades that could be done safely, and worked on security advisories, which will get published as part of the next release
   - We modernized and simplified our ancient scripts used to [build our deb packages](https://github.com/ckan/ckan-packaging/pull/24)

* A bit less activity on the DCAT front because @amercader was focusing on 2.11 but still manage to finish the [SHACL validation integration](https://github.com/ckan/ckanext-dcat/pull/288) and [support for classes/ranges](https://github.com/ckan/ckanext-dcat/pull/291)

* @wardi still finding more parts of CKAN to make faster, with a focus on dataset updates performance (https://github.com/ckan/ckan/pull/8360) with some big gains for `package_update`-related APIs, as well as working towards a new and better `ckan db init` command (https://github.com/ckan/ckan/pull/8336)

**Comment by amercader** [comment]
## August 2024

Some of us took a couple weeks break to recharge our batteries but that didn't prevent a lot of important stuff happening during August!

* :ringed_planet: :comet: **CKAN 2.11** was finally released with a ton of good stuff. Check out the [announcement post](https://ckan.org/blog/the-new-ckan-211-release-is-now-available) and the massive [CHANGELOG](https://docs.ckan.org/en/2.11/changelog.html#v-2-11-0-2024-08-21). As ever, the release involved several tasks on many fronts, and we took the opportunity to build automation around some of them, which we'll consolidate and put in place in the coming months to make future releases smoother and more frequent.
* A new major version of **ckanext-dcat** was also released, 2.0.0. This includes all recent work about ckanext-scheming support and an initial profile for **DCAT AP v3**. Testing and feedback around these and the other new features is greatly appreciated. Check out the [CHANGELOG](https://docs.ckan.org/projects/ckanext-dcat/en/latest/changelog/#v200-2024-08-30), now included in a revised and updated documentation: https://docs.ckan.org/projects/ckanext-dcat/
* But work doesn't stop and improvements keep coming, @wardi worked internals in https://github.com/ckan/ckan/pull/8396 and continues focusing on package_update , working on allowing to edit metadata_modified and omitting the resource list. Both he and @smotornyuk will focus on integrating a new file management system.
@kowh-ai keeps working on the Docker images to improve them and make them easier to use

**Comment by amercader** [comment]
## September 2024

A busy September with no major milestones to report but lots of smaller contributions on different fronts:

* @wardi worked on yet another performance improving change, in this case delivering massive speedups on datasets with many resources. The PR is pending review (https://github.com/ckan/ckan/pull/8421) but should be merged soon, ready for the next CKAN release. This change also allows omitting the resources list to `package_update`, setting the `metadata_modified` field for sysadmins and prevents activities from being created when no actual change occurred.
* We started discussing important work around packaging and requirements management, see https://github.com/ckan/ckan/issues/8382, https://github.com/ckan/ckan/issues/8465 and https://github.com/ckan/ckan/pull/8456. There are multiple interconnected issues that need to be juggled to make sure we balance moving forward and keeping compatibility.
* @wardi and @smotornyuk are drafting a plan to revamp the file management in CKAN to allow things like linking files to arbitrary entities, multiple backend support, efficient handling of huge files, etc. Stay tuned for more news on this.
* We've been imagining how a new search interface with support for multiple backends could look like (https://github.com/ckan/ckan/discussions/8444), and supporting the ongoing efforts on the new frontend.

**Comment by amercader** [comment]
## October 2024

October came and went with lots of activity in different areas. No major announcements but lot of small high value bits:

* There are new releases for ckanext-dcat and ckanext-harvest. [Ckanext-dcat 2.1.0](https://docs.ckan.org/projects/ckanext-dcat/en/latest/changelog/#v210-2024-10-31) introduces [multilingual support](https://docs.ckan.org/projects/ckanext-dcat/en/latest/profiles/#multilingual-support) and a new base profile for the [DCAT US v3 spec](https://doi-do.github.io/dcat-us/), while [ckanext-harvest 1.6.0](https://github.com/ckan/ckanext-harvest/releases/tag/v1.6.0) contains various fixes for CKAN 2.11 support.
* @wardi started work on a preview for the new file uploads work, we'll keep you posted as this progressed
* Both @wardi and myself attended conferences to showcase CKAN (at the [Canadian Open Data Summit](https://www.opendatasummit.ca/event/cods24scdo24/schedule/SESEYHMZ2CG0JQ4FD) and the [CLARIN Annual Conference](https://www.clarin.eu/event/2024/building-bridges-industry-clarin2024)) so we spent a bit of time preparing for that
* We provided support on the ongoing theme refreshment work led by @aleeexgreeen , as well as various pull requests and issues.

**Comment by amercader** [comment]
## November 2024

Another busy month in CKAN land:

* A new release was made for the official CKAN Docker images, that includes changes in user permissions to make the images more secure. We also officially deprecated the Alpine based images. Both these two might imply changes on your GitHub Actions workflow files. Check the [release notes](https://github.com/ckan/ckan-docker-base/releases/tag/v20241125) for all details.
* A new release for ckanext-spatial: [v2.2.0](https://docs.ckan.org/projects/ckanext-spatial/en/latest/changelog/#v220-2024-11-08) fixes various bugs and expands support for CKAN 2.11 and SQLAlchemy 2. We also reviewed and updated all the [documentation](https://docs.ckan.org/projects/ckanext-spatial/en/latest/), check it out!
* We have started preparing a new patch release for CKAN 2.11 that includes various fixes. We'll take this opportunity to review the release process and deploy automations to make the process simpler and more secure (e.g. automatic [publishing to PyPI](https://github.com/ckan/ckan/pull/8520) and [Docker Hub](https://github.com/ckan/ckan-docker-base/pull/73))
* On the DataStore front, new default datastore settings were added to save approximately 50% database storage space and significantly improve loading time at a small potential performance cost for FTS search on specific fields. See [#8530](https://github.com/ckan/ckan/pull/8530) for more details. This setting will also be available on CKAN 2.11.1
* Work on the new theme is humming along, and well as the usual management of incoming PRs and issues.

**Comment by amercader** [comment]
## December 2024 / January 2025

The end of last year was so busy we forgot to post the monthly update before going on holidays, but we'll make up for that with a double-sized update this time!

* In December we published the 2.11.1 and 2.10.6 patch releases and took the opportunity to review and update the whole [release process](https://docs.ckan.org/en/latest/contributing/release-process.html), streamlining it and automating all the publishing processes to increase efficiency and security. Apart from the tasks mentioned in previous updates, we finished the automatic publishing of [deb packages](https://github.com/ckan/ckan-packaging/pull/24) and [GitHub releases](https://github.com/ckan/ckan/pull/8570).

* We've been working on some security enhancements that will be released in yet another patch release [next week](https://groups.google.com/a/ckan.org/g/ckan-announce/c/QnMsWTJ3OVc), so stay tuned.

* @wardi worked on improving the performance when dealing with large data files. Changes like [#8590](https://github.com/ckan/ckan/pull/8590), [#8597](https://github.com/ckan/ckan/pull/8597) and [#8603](https://github.com/ckan/ckan/pull/8603) can make rendering the resource page ~20x faster and the data table view ~100x faster for some large resources. Work started towards further improvements on filtering data tables as well.

* Work is steadily progressing in the new theme front with [multiple PRs](https://github.com/ckan/ckan/pulls?q=is%3Apr+author%3Aaleeexgreeen+is%3Aclosed+) by @aleeexgreeen who is stoically addressing the many review comments.

* Lots of activity in the metadata formats side of things. [ckanext-dcat 2.2.0](https://docs.ckan.org/projects/ckanext-dcat/en/latest/changelog/#v220-2025-01-30) was just released and includes the first community contributed profile, for [HealthDCAT-AP](https://healthdcat-ap.github.io/), thanks to folks at Health-RI. There's another ongoing effort to include support for the [Croissant ML](https://docs.mlcommons.org/croissant/) specification, which will make CKAN a first class choice for hosting ML-ready data. Work is nearing completion [here](https://github.com/ckan/ckanext-dcat/pull/330) and we'll provide more details when it's finished and published.

* A new extension ([ckanext-dataset-series](https://github.com/ckan/ckanext-dataset-series)) was created to provide a fast and simple implementation for [Dataset Series](https://github.com/ckan/ckanext-dcat/issues/298). These are collections of datasets than can be ordered, i.e. allowing navigating to previous and next items. If you are interested check the linked issue and give the extension a try to provide feedback!

**Comment by amercader** [comment]
## February 2025

Here is yet another late update, but rest assured that work doesn't stop on our end

* The CKAN 2.10.7 and 2.11.2 patch releases were released at the [beginning of the month](https://ckan.org/blog/new-patch-releases-for-ckan-211x-and-210x-are-now-available). If you missed it now is the time to upgrade your instances.
* The latest ckanext-dcat release ([2.3.0](https://docs.ckan.org/projects/ckanext-dcat/en/latest/changelog/#v230-2025-02-25)) contains a new profile to support the [Croissant](https://github.com/mlcommons/croissant) format, a community standard to describe Machine Learning datasets, that will allow CKAN to integrate with many tools in the ML/AI ecosystem. A blog post with more details in ckan.org is imminent so stay tuned.
* @wardi has been working on preliminary work to import/export data schemas with table designer and tying up some earlier DataTables performance work
* Lots of review work to get PRs merged around new theme, caching, authentication, etc

**Comment by amercader** [comment]
## March 2025

The Spring Equinox went by, days are getting shorter or longer depending on what hemisphere you are in, but work never stops at the CKAN workshop:

* Early in the month we released a new version of the [Solr Docker images](https://github.com/ckan/ckan-solr/releases/tag/v20250303) to include some upstream security fixes.
* A new [ckanext-scheming](https://github.com/ckan/ckanext-scheming/blob/master/CHANGELOG.md#310) release is out with loads of fixes and new features.
* A small [ckanext-spatial](https://github.com/ckan/ckanext-spatial/releases/tag/v2.3.0) release also went to out to address some requirements issue and bug fixes.
* We started work on a proof of concept for the longly-discussed [new search](https://github.com/ckan/ckan/discussions/8444) feature. This is an ambitious rethinking of CKAN's
  current search capabilities: pluggable backends, support for multiple entities, automatic fields indexing... There's a long road ahead but the first results seem really  promising. We'll update when work has progressed more but if you want to follow along, development is being done [in this repo](https://github.com/amercader/ckanext-search).
* We are finally seeing the initial stages for the integration into CKAN of @smotornyuk's great work on a new file management infrastructure ([#8745](https://github.com/ckan/ckan/pull/8745))
* Most of the DataStore / DataTables performance PRs by @wardi have now been updated and merged, see e.g. #[8590](https://github.com/ckan/ckan/pull/8590) and [#8603](https://github.com/ckan/ckan/pull/8603)
* Big pull requests reviewed and merged, e.g. around [simplifying dataset indexing](https://github.com/ckan/ckan/pull/8396), [removing the model](https://github.com/ckan/ckan/pull/8702) from the context object, and [dashboard](https://github.com/ckan/ckan/pull/8675) and [dataset form](https://github.com/ckan/ckan/pull/8503) pages for the new theme.

**Comment by amercader** [comment]
## April 2025

Yet another late update, but lots of cool stuff happening last month:

* Thanks to @duttonw we have a new CI setup based on GitHub Actions that improves performance and tests coverage. We worked on consolidating all checks across release branches and improving the overall workflow, benefiting all contributors with faster and clearer feedback(#8744, #8783)
* @smotornyuk continues to flesh out the next version of CKAN's file management capabilities, now with a clear roadmap of the different phases that the process will take: https://github.com/ckan/ckan/issues/8920
* Work also continues on the new search feature, currently focusing on defining a reusable [Query filters specification](https://github.com/amercader/ckanext-search/pull/1) as well as support for it in different backends
* @wardi has been focusing on integrating Htmx in different parts of the UI to improve performance and user experience, see e.g. https://github.com/ckan/ckan/pull/8808
* Lots of review work to get PRs merged as well maintenance work like upgrading requirements, improvements and fixes in the Docker images, etc

**Comment by amercader** [comment]
## May 2025

Here's what we were up to this month:

 * We started the month strong with new patch releases: [2.11.3](https://docs.ckan.org/en/2.11/changelog.html#v-2-11-3-2025-05-07) and [2.10.8](https://docs.ckan.org/en/2.10/changelog.html#v-2-10-8-2025-05-07). Update your sites to get loads of bug fixes and minor fixes.
 * [ckanext-dcat 2.4.0](https://docs.ckan.org/projects/ckanext-dcat/en/latest/changelog/#v240-2025-05-20) introduced support for [Dataset Series](https://docs.ckan.org/projects/ckanext-dcat/en/latest/dataset-series/) via integration with [ckanext-dataset-series](https://github.com/ckan/ckanext-dataset-series). Ckanext-dataset-series takes care of all the underlying management of the series, and ckanext-dcat exposes them using `dcat:DatasetSeries` classes. Check the [documentation](https://docs.ckan.org/projects/ckanext-dcat/en/latest/dataset-series/) for all details on how to set them up
 * @wardi carried on improving the user experience on the frontend, this time with the dataset search page, one of the most important and complex pages in CKAN. Props to @mutantsan for great support on reviewing https://github.com/ckan/ckan/pull/8930
 * Work on search query filters is nearing completion (https://github.com/amercader/ckanext-search/pull/1) and the plan is to test drive it in the wider search POC and potentially the current DataStore search action.
 * Work progresses steadily on the new frontend, with @aleeexgreeen more recently focusing on important pages like the org/group (https://github.com/ckan/ckan/pull/8923) and dataset/resources (https://github.com/ckan/ckan/pull/8965) ones

**Comment by amercader** [comment]
## June 2025

A quiet month, but still some cool developments under way:
* @wardi started thinking of ways to add a proper model for [Data Dictionaries](https://github.com/ckan/ckan/discussions/8531#discussioncomment-13435947) and persist them in order to unlock a whole new set of features. Leave your comments in the discussion above if you are interested in this topic.
* The new Query Filters implementation was successfully integrated in the search providers for the new search POC. Next up is finish the rest of input parameters like pagination, facets, etc.
* Lots of grunt work related to [requirements upgrades](https://github.com/ckan/ckan/pull/8995), [version changes](https://github.com/ckan/ckan/pull/8998) and [deprecations](https://github.com/ckan/ckan/pull/8992).
